src/mpid/ch4/netmod/ofi/init_settings.c

```diff
@@ -0,0 +1,357 @@
+    }
+
+    /* With scalable endpoints, FI_NAMED_RX_CTX is needed to specify a destination receive context
+     * index */
+    if (MPIDI_OFI_ENABLE_SCALABLE_ENDPOINTS)
+        hints->caps |= FI_NAMED_RX_CTX;
+
+    /* ------------------------------------------------------------------------ */
+    /* Set object options to be filtered by getinfo                             */
+    /* domain_attr:  domain attribute requirements                              */
+    /* op_flags:     persistent flag settings for an endpoint                   */
+    /* endpoint type:  see FI_EP_RDM                                            */
+    /* Filters applied (for this netmod, we need providers that can support):   */
+    /* THREAD_DOMAIN:  Progress serialization is handled by netmod (locking)    */
+    /* PROGRESS_AUTO:  request providers that make progress without requiring   */
+    /*                 the ADI to dedicate a thread to advance the state        */
+    /* FI_DELIVERY_COMPLETE:  RMA operations are visible in remote memory       */
+    /* FI_COMPLETION:  Selective completions of RMA ops                         */
+    /* FI_EP_RDM:  Reliable datagram                                            */
+    /* ------------------------------------------------------------------------ */
+    hints->addr_format = FI_FORMAT_UNSPEC;
+    hints->domain_attr->threading = FI_THREAD_DOMAIN;
+    MPIDI_OFI_set_auto_progress(hints);
+    hints->domain_attr->resource_mgmt = FI_RM_ENABLED;
+    hints->domain_attr->av_type = MPIDI_OFI_ENABLE_AV_TABLE ? FI_AV_TABLE : FI_AV_MAP;
+
+    if (fi_version() >= FI_VERSION(1, 5)) {
+        hints->domain_attr->mr_mode = 0;
+#ifdef FI_RESTRICTED_COMP
+        hints->domain_attr->mode = FI_RESTRICTED_COMP;
+#endif
+
+#ifdef FI_MR_VIRT_ADDR
+        if (MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS) {
+            hints->domain_attr->mr_mode |= FI_MR_VIRT_ADDR;
+        }
+#endif
+
+#ifdef FI_MR_ALLOCATED
+        if (MPIDI_OFI_ENABLE_MR_ALLOCATED) {
+            hints->domain_attr->mr_mode |= FI_MR_ALLOCATED;
+        }
+#endif
+
+#ifdef FI_MR_PROV_KEY
+        if (MPIDI_OFI_ENABLE_MR_PROV_KEY) {
+            hints->domain_attr->mr_mode |= FI_MR_PROV_KEY;
+        }
+#endif
+    } else {
+        /* In old versions FI_MR_BASIC is equivallent to set
+         * FI_MR_VIRT_ADDR, FI_MR_PROV_KEY, and FI_MR_ALLOCATED on.
+         * FI_MR_SCALABLE is equivallent to all bits off in newer versions.
+         */
+        MPIR_Assert(MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS == MPIDI_OFI_ENABLE_MR_PROV_KEY);
+        if (MPIDI_OFI_ENABLE_MR_VIRT_ADDRESS) {
+            hints->domain_attr->mr_mode = FI_MR_BASIC;
+        } else {
+            hints->domain_attr->mr_mode = FI_MR_SCALABLE;
+        }
+    }
+    /* FI_MR_SCALABLE is implied by lack of mr_mode bits in >= v1.5.
+     * But at least sockets provider still need FI_MR_SCALABLE to be set
+     * or it will return FI_MR_BASIC. */
```

Comment:
I'm not sure that's true. What needs to happen isn't about setting `FI_MR_SCALABLE`, it's about clearing any other mode bits.

Reply:
For sockets provider, if you clear all mode bits, it will default to `FI_MR_BASIC` -- opposite to the intention.

Reply:
I can't see where the sockets provider uses the mode bits at all honestly. I see a bunch of places in where it sets the bits, but I don't see anywhere that they're actually used. The only place I see that seems to touch this again is in `ofi_mr_map_init()`. In that spot, if the mode bits are all unset, it sets the mode to `FI_MR_SCALABLE`.

https://github.com/ofiwg/libfabric/blob/14b3cd0450aebe8b74a2c5ed219a53cac06d3e0a/prov/util/src/util_mr_map.c#L164-L189